tpe/plugin/src/SDFFeatures.cc

```diff
@@ -0,0 +1,222 @@
+  }
+
+  return modelIdentity;
+}
+
+/////////////////////////////////////////////////
+Identity SDFFeatures::ConstructSdfLink(
+    const Identity &_modelID,
+    const ::sdf::Link &_sdfLink)
+{
+  // Read sdf params
+  const std::string name = _sdfLink.Name();
+  igndbg << "construct link " << name << std::endl;
+  const auto pose = _sdfLink.RawPose();
+
+  auto it = this->models.find(_modelID);
+  if (it == this->models.end())
+  {
+    ignwarn << "Model [" << _modelID.id << "] is not found" << std::endl;
+    return this->GenerateInvalidId();
+  } 
+  auto model = it->second->model;
+  if (model == nullptr)
+  {
+    ignwarn << "Model is a nullptr" << std::endl;
+    return this->GenerateInvalidId();
+  }
+  tpelib::Entity &ent = model->AddLink();
+  tpelib::Link *link = static_cast<tpelib::Link *>(&ent);
+  link->SetName(name);
+  link->SetPose(pose);
+  const auto linkIdentity = this->AddLink(model->GetId(), *link);
+
+  // construct collisions
+  for (std::size_t i = 0; i < _sdfLink.CollisionCount(); ++i)
+  {
+    this->ConstructSdfCollision(linkIdentity, *_sdfLink.CollisionByIndex(i));
+  }
+
+  return linkIdentity;
+}
+
+/////////////////////////////////////////////////
+Identity SDFFeatures::ConstructSdfCollision(
+    const Identity &_linkID,
+    const ::sdf::Collision &_sdfCollision)
+{
+  // Read sdf params
+  const std::string name = _sdfCollision.Name();
+  igndbg << "construct collision " << name << std::endl;
+  const auto pose = _sdfCollision.RawPose();
+  const auto geom = _sdfCollision.Geom();
+
+  auto it = this->links.find(_linkID);
+  if (it == this->links.end())
+  {
+    ignwarn << "Link [" << _linkID.id << "] is not found" << std::endl;
+    return this->GenerateInvalidId();
+  }
+  auto link = it->second->link;
+  if (link == nullptr)
+  {
+    ignwarn << "Link is a nullptr" << std::endl;
+    return this->GenerateInvalidId();
+  }
+
+  tpelib::Entity &ent = link->AddCollision();
+  tpelib::Collision *collision = static_cast<tpelib::Collision *>(&ent);
+  collision->SetName(name);
+  collision->SetPose(pose);
+  if (geom->Type() == ::sdf::GeometryType::BOX)
+  {
+    const auto boxSdf = geom->BoxShape();
+    tpelib::BoxShape shape;
+    shape.SetSize(boxSdf->Size());
+    collision->SetShape(shape);
+  }
+  else if (geom->Type() == ::sdf::GeometryType::CYLINDER)
+  {
+    const auto cylinderSdf = geom->CylinderShape();
+    tpelib::CylinderShape shape;
+    shape.SetRadius(cylinderSdf->Radius());
+    shape.SetLength(cylinderSdf->Length());
+    collision->SetShape(shape);
+  }
+  else if (geom->Type() == ::sdf::GeometryType::SPHERE)
+  {
+    const auto sphereSdf = geom->SphereShape();
+    tpelib::SphereShape shape;
+    shape.SetRadius(sphereSdf->Radius());
+    collision->SetShape(shape);
+  }
+  const auto collisionIdentity = this->AddCollision(link->GetId(), *collision);
+  return collisionIdentity;
+}
+
+/////////////////////////////////////////////////
+Identity SDFFeatures::ConstructSdfCollision(
```

Comment:
I'd like to think there's function overloading happening, but you might be right. The problem here is that mesh has to be passed in as a parameter and doing so violates the `ConstructSdfCollision` [function template](https://github.com/ignitionrobotics/ign-physics/blob/master/sdf/include/ignition/physics/sdf/ConstructCollision.hh#L42) I'm not sure what's the best thing to do here. I guess I can remove the function for now if the support for mesh is not definite. @iche033 what do you think?

Reply:
I think you could add an `else if (geom->Type() == ::sdf::GeometryType::MESH)` to the other `ConstructSdfCollision`. Then the question is how to go from `_sdfCollision->MeshShape()` to `common::Mesh`.

Or if meshes are not an important use case for TPE, this could be removed for now.

Reply:
it may be fine to remove it. I may have added it think that it's necessary but `dartsim` also does not implement this. We can always add it back if needed

Reply:
actually `dartsim` does have the `ConstructSdfCollision` function implemented but it's just not used by ign-gazebo. There is also a [TODO note](https://github.com/ignitionrobotics/ign-physics/blob/master/dartsim/src/SDFFeatures.cc#L291) for constructing meshes so I'm not sure if that function works for meshes or not.

We could either remove ConstructSdfCollision for tpe or just do what dartsim does and ignore meshes for now.

Reply:
I did what dartsim does and removed the function for meshes https://github.com/ignitionrobotics/ign-physics/pull/46/commits/8f58c69cfdae5c0e0d83f3f3c99e9a2cc349d939 also kept the TODO

Reply:
lgtm